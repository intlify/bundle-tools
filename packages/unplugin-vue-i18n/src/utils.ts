import { promises as fs } from 'fs'
import pc from 'picocolors'

export function warn(...args: unknown[]) {
  console.warn(pc.yellow(pc.bold(`[unplugin-vue-i18n] `)), ...args)
}

export function error(...args: unknown[]) {
  console.error(pc.red(pc.bold(`[unplugin-vue-i18n] `)), ...args)
}

export async function getRaw(path: string): Promise<string> {
  return fs.readFile(path, { encoding: 'utf-8' })
}

export function raiseError(message: string) {
  throw new Error(`[unplugin-vue-i18n] ${message}`)
}
