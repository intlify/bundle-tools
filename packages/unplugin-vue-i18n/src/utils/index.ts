export * from './log'
export * from './plugin'
export * from './resolver'
