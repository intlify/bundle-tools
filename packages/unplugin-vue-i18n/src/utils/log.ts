import pc from 'picocolors'
import { PKG_NAME } from '../constants'

export function warn(...args: unknown[]) {
  console.warn(pc.yellow(pc.bold(`[${PKG_NAME}] `)), ...args)
}

export function error(...args: unknown[]) {
  console.error(pc.red(pc.bold(`[${PKG_NAME}] `)), ...args)
}

export function raiseError(message: string) {
  throw new Error(`[${PKG_NAME}] ${message}`)
}
