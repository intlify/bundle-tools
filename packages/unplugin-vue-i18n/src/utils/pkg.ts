// SPDX-License-Identifier: MIT
// Modified by: kazuya kawaguchi (a.k.a. kazupon)
// Auther: Pooya Parsa (https://github.com/pi0)
// Forked from: https://github.com/unjs/pkg-types
// Note: Modified to work as a synchronous API

import { statSync } from 'node:fs'
import { join, resolve, isAbsolute } from 'pathe'
import { resolvePathSync } from 'mlly'

import type { ResolveOptions as _ResolveOptions } from 'mlly'

export interface FindFileOptions {
  /**
   * The starting directory for the search.
   * @default . (same as `process.cwd()`)
   */
  startingFrom?: string
  /**
   * A pattern to match a path segment above which you don't want to ascend
   * @default /^node_modules$/
   */
  rootPattern?: RegExp
  /**
   * If true, search starts from root level descending into subdirectories
   */
  reverse?: boolean
  /**
   * A matcher that can evaluate whether the given path is a valid file (for example,
   * by testing whether the file path exists.
   *
   * @default fs.statSync(path).isFile()
   */
  test?: (filePath: string) => boolean | undefined
}

const defaultFindOptions: Required<FindFileOptions> = {
  startingFrom: '.',
  rootPattern: /^node_modules$/,
  reverse: false,
  test: (filePath: string) => {
    try {
      if (statSync(filePath).isFile()) {
        return true
      }
    } catch {
      // Ignore
    }
  }
}

/**
 * Asynchronously finds a file by name, starting from the specified directory and traversing up (or down if reverse).
 * @param filename - The name of the file to find.
 * @param _options - Options to customise the search behaviour.
 * @returns a promise that resolves to the path of the file found.
 * @throws Will throw an error if the file cannot be found.
 */
function findFile(
  filename: string | string[],
  _options: FindFileOptions = {}
): string {
  const filenames = Array.isArray(filename) ? filename : [filename]
  const options = { ...defaultFindOptions, ..._options }
  const basePath = resolve(options.startingFrom)
  const leadingSlash = basePath[0] === '/'
  const segments = basePath.split('/').filter(Boolean)

  // Restore leading slash
  if (leadingSlash) {
    segments[0] = '/' + segments[0]
  }

  // Limit to node_modules scope if it exists
  let root = segments.findIndex(r => r.match(options.rootPattern))
  if (root === -1) {
    root = 0
  }

  if (options.reverse) {
    for (let index = root + 1; index <= segments.length; index++) {
      for (const filename of filenames) {
        const filePath = join(...segments.slice(0, index), filename)
        if (options.test(filePath)) {
          return filePath
        }
      }
    }
  } else {
    for (let index = segments.length; index > root; index--) {
      for (const filename of filenames) {
        const filePath = join(...segments.slice(0, index), filename)
        if (options.test(filePath)) {
          return filePath
        }
      }
    }
  }

  throw new Error(
    `Cannot find matching ${filename} in ${options.startingFrom} or parent directories`
  )
}

/**
 * Asynchronously finds the next file with the given name, starting in the given directory and moving up.
 * Alias for findFile without reversing the search.
 * @param filename - The name of the file to find.
 * @param _options - Options to customise the search behaviour.
 * @returns A promise that resolves to the path of the next file found.
 */
function findNearestFile(
  filename: string | string[],
  _options: FindFileOptions = {}
): string {
  return findFile(filename, _options)
}

/**
 * Represents the options for resolving paths with additional file finding capabilities.
 */
export type ResolveOptions = _ResolveOptions & FindFileOptions

/**
 * Resolves the path to the nearest `package.json` file from a given directory.
 * @param id - The base path for the search, defaults to the current working directory.
 * @param options - Options to modify the search behaviour. See {@link ResolveOptions}.
 * @returns A promise resolving to the path of the nearest `package.json` file.
 */
export function resolvePackageJSON(
  id: string = process.cwd(),
  options: ResolveOptions = {}
): string {
  const resolvedPath = isAbsolute(id) ? id : resolvePathSync(id, options)
  return findNearestFile('package.json', {
    startingFrom: resolvedPath,
    ...options
  })
}
