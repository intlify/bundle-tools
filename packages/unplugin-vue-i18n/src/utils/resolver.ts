import fs from 'node:fs'
import createDebug from 'debug'
import module from 'node:module'
import path from 'node:path'

const SUPPORT_PACKAGES = ['vue-i18n', 'petite-vue-i18n'] as const

type SupportPackage = (typeof SUPPORT_PACKAGES)[number]

export type InstalledPackageInfo = {
  alias: string
  pkg: SupportPackage
}

const _require = module.createRequire(import.meta.url)

export function checkInstallPackage(
  debug: createDebug.Debugger
): InstalledPackageInfo {
  const pkgInfo =
    resolvePkgPath('vue-i18n', debug) ||
    resolvePkgPath('petite-vue-i18n', debug)
  if (!pkgInfo) {
    throw new Error(
      `requires 'vue-i18n' or 'petite-vue-i18n' to be present in the dependency tree.`
    )
  }

  debug('installed package info:', pkgInfo)
  return pkgInfo
}

function resolvePkgPath(
  id: string,
  debug: createDebug.Debugger
): InstalledPackageInfo | null {
  try {
    /**
     * NOTE:
     *  Assuming the case of using npm alias `npm:`,
     *  get the installed package name from `package.json`
     */
    const resolvedPath = _require.resolve(id)
    const pkgPath = path.dirname(resolvedPath)
    const pkgJson = JSON.parse(
      fs.readFileSync(path.join(pkgPath, 'package.json'), 'utf-8')
    ) as { name: string }
    const pkgName: string = pkgJson.name.startsWith('vue-i18n')
      ? 'vue-i18n'
      : pkgJson.name.startsWith('petite-vue-i18n')
        ? 'petite-vue-i18n'
        : ''
    return pkgJson ? { alias: id, pkg: pkgName as SupportPackage } : null
  } catch (e) {
    debug(`cannot find '${id}'`, e)
    return null
  }
}
