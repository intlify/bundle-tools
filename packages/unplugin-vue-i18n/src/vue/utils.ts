import type * as _compiler from 'vue/compiler-sfc'
import type { RollupPlugin } from 'unplugin'

export function getVueCompiler(vuePlugin: RollupPlugin): typeof _compiler {
  return vuePlugin?.api?.options.compiler as typeof _compiler
}

export type VuePluginResolvedOptions = {
  isProduction: boolean
  root: string
  compiler: typeof _compiler
  template?: Partial<
    Omit<
      _compiler.SFCTemplateCompileOptions,
      | 'id'
      | 'source'
      | 'ast'
      | 'filename'
      | 'scoped'
      | 'slotted'
      | 'isProd'
      | 'inMap'
      | 'ssr'
      | 'ssrCssVars'
      | 'preprocessLang'
    >
  >
}

export function getVuePluginOptions(
  vuePlugin: RollupPlugin
): VuePluginResolvedOptions {
  return {
    isProduction: vuePlugin?.api?.options.isProduction,
    root: vuePlugin?.api?.options.root,
    template: vuePlugin?.api?.options.template,
    compiler: vuePlugin?.api?.options.compiler as typeof _compiler
  }
}
