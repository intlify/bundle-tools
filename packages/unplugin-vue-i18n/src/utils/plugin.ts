import path from 'node:path'
import { PKG_NAME } from '../constants'
import { error } from './log'

import type { VitePlugin, RollupPlugin } from 'unplugin'

export function resolveNamespace(name: string): string {
  return `${PKG_NAME}:${name}`
}

// @ts-expect-error -- FIXME: plugin type
type UserConfig = Parameters<VitePlugin['configResolved']>[0]

export function getVitePlugin(
  config: UserConfig,
  name: string
): RollupPlugin | null {
  // vite plugin has compoaibility for rollup plugin
  return config.plugins.find(p => p.name === name) as RollupPlugin
}

export function checkVuePlugin(vuePlugin: RollupPlugin | null): boolean {
  if (vuePlugin == null || !vuePlugin.api) {
    error(
      '`@vitejs/plugin-vue` plugin is not found or invalid version. Please install `@vitejs/plugin-vue` v4.3.4 or later version.'
    )
    return false
  }
  return true
}

const isWindows = typeof process !== 'undefined' && process.platform === 'win32'

const windowsSlashRE = /\\/g
function slash(p: string): string {
  return p.replace(windowsSlashRE, '/')
}

export function normalizePath(id: string): string {
  return path.posix.normalize(isWindows ? slash(id) : id)
}
