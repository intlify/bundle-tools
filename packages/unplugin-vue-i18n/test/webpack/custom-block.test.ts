import { bundleWebpack, bundleAndRun } from '../utils'
import { createMessageContext, compile } from '@intlify/core-base'
import type { MessageCompilerContext } from '@intlify/core-base'

test('basic', async () => {
  const { module } = await bundleAndRun('basic.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()

  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  const fn = compile(i18n.resource.en.hello, {} as MessageCompilerContext)
  expect(fn(createMessageContext())).toEqual('hello world!')
})

test('yaml', async () => {
  const { module } = await bundleAndRun('yaml.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()

  let i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('ja')
  const fn1 = compile(i18n.resource.hello, {
    locale: i18n.locale
  } as MessageCompilerContext)
  expect(fn1(createMessageContext())).toEqual('こんにちは、世界！')

  i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('en')
  const fn2 = compile(i18n.resource.hello, {
    locale: i18n.locale
  } as MessageCompilerContext)
  expect(fn2(createMessageContext())).toEqual('hello world!')
})

test('json5', async () => {
  const { module } = await bundleAndRun('json5.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()

  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  const fn = compile(i18n.resource.en.hello, {} as MessageCompilerContext)
  expect(fn(createMessageContext())).toEqual('hello world!')
})

test('import', async () => {
  const { module } = await bundleAndRun('import.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()

  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  const fn = compile(i18n.resource.en.hello, {} as MessageCompilerContext)
  expect(fn(createMessageContext())).toEqual('hello world!')
})

test('multiple', async () => {
  const { module } = await bundleAndRun('multiple.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()
  expect(module.__i18n.length).toEqual(2)

  let i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  const fn1 = compile(i18n.resource.ja.hello, {} as MessageCompilerContext)
  expect(fn1(createMessageContext())).toEqual('こんにちは、世界！')

  i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  const fn2 = compile(i18n.resource.en.hello, {} as MessageCompilerContext)
  expect(fn2(createMessageContext())).toEqual('hello world!')
})

test('locale attr', async () => {
  const { module } = await bundleAndRun('locale.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()

  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('ja')
  const fn = compile(i18n.resource.hello, {
    locale: i18n.locale
  } as MessageCompilerContext)
  expect(fn(createMessageContext())).toEqual('こんにちは、世界！')
})

test('locale attr and basic', async () => {
  const { module } = await bundleAndRun('locale-mix.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()

  let i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('ja')
  const fn1 = compile(i18n.resource.hello, {
    locale: i18n.locale
  } as MessageCompilerContext)
  expect(fn1(createMessageContext())).toEqual('こんにちは、世界！')

  i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  const fn2 = compile(i18n.resource.en.hello, {} as MessageCompilerContext)
  expect(fn2(createMessageContext())).toEqual('hello world!')
})

test('locale attr and import', async () => {
  const { module } = await bundleAndRun('locale-import.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()

  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('en')
  const fn = compile(i18n.resource.hello, {
    locale: i18n.locale
  } as MessageCompilerContext)
  expect(fn(createMessageContext())).toEqual('hello world!')
})

test('special characters', async () => {
  const { module } = await bundleAndRun('special-char.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()

  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  const fn = compile(i18n.resource.en.hello, {} as MessageCompilerContext)
  expect(fn(createMessageContext())).toEqual('hello\ngreat\t"world"')
})

test('global', async () => {
  const { module } = await bundleAndRun('global-mix.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()
  expect(module.__i18nGlobal).toMatchSnapshot()

  const l = module.__i18n.pop()
  expect(l.locale).toEqual('ja')
  const fn1 = compile(l.resource.hello, {
    locale: l.locale
  } as MessageCompilerContext)
  expect(fn1(createMessageContext())).toEqual('hello local!')

  const g = module.__i18nGlobal.pop()
  expect(g.locale).toEqual('')
  const fn2 = compile(g.resource.en.hello, {
    locale: g.locale
  } as MessageCompilerContext)
  expect(fn2(createMessageContext())).toEqual('hello global!')
})

test('default lang', async () => {
  const { module } = await bundleAndRun('default-lang.vue', bundleWebpack, {
    defaultSFCLang: 'yml'
  })
  expect(module.__i18n).toMatchSnapshot()
  const l = module.__i18n.pop()
  const fn = compile(l.resource.en.hello, {} as MessageCompilerContext)
  expect(fn(createMessageContext())).toEqual('hello from defaults!')
})

test('default lang and global scope', async () => {
  const { module } = await bundleAndRun('default-lang.vue', bundleWebpack, {
    defaultSFCLang: 'yml',
    globalSFCScope: true
  })
  expect(module.__i18nGlobal).toMatchSnapshot()
  const g = module.__i18nGlobal.pop()
  const fn = compile(g.resource.en.hello, {} as MessageCompilerContext)
  expect(fn(createMessageContext())).toEqual('hello from defaults!')
})

test('global scope and import', async () => {
  const { module } = await bundleAndRun(
    'global-scope-import.vue',
    bundleWebpack,
    {
      globalSFCScope: true
    }
  )
  expect(module.__i18nGlobal).toMatchSnapshot()
  const g = module.__i18nGlobal.pop()
  const fn = compile(g.resource.en.hello, {} as MessageCompilerContext)
  expect(fn(createMessageContext())).toEqual('hello world!')
})

test('array', async () => {
  const { module } = await bundleAndRun('array.vue', bundleWebpack)
  expect(module.__i18n).toMatchSnapshot()

  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  const fn1 = compile(i18n.resource.en.foo[0][0], {} as MessageCompilerContext)
  const fn2 = compile(i18n.resource.en.foo[1], {} as MessageCompilerContext)
  expect(fn1(createMessageContext())).toEqual('bar')
  expect(fn2(createMessageContext())).toEqual('baz')
})
