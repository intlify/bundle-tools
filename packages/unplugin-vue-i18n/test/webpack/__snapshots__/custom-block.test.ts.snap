// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`array 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "foo": [
          [
            [Function],
          ],
          [Function],
        ],
      },
    },
  },
]
`;

exports[`basic 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`default lang 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`default lang and global scope 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`global 1`] = `
[
  {
    "locale": "ja",
    "resource": {
      "hello": [Function],
    },
  },
]
`;

exports[`global 2`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`global scope and import 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`import 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`json5 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`locale attr 1`] = `
[
  {
    "locale": "ja",
    "resource": {
      "hello": [Function],
    },
  },
]
`;

exports[`locale attr and basic 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
  {
    "locale": "ja",
    "resource": {
      "hello": [Function],
    },
  },
]
`;

exports[`locale attr and import 1`] = `
[
  {
    "locale": "en",
    "resource": {
      "hello": [Function],
    },
  },
]
`;

exports[`multiple 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
  {
    "locale": "",
    "resource": {
      "ja": {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`special characters 1`] = `
[
  {
    "locale": "",
    "resource": {
      "en": {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`yaml 1`] = `
[
  {
    "locale": "en",
    "resource": {
      "hello": [Function],
    },
  },
  {
    "locale": "ja",
    "resource": {
      "hello": [Function],
    },
  },
]
`;
